src/Microsoft.AspNetCore.SignalR.Core/HubConnectionContext.cs

```diff
@@ -156,27 +154,32 @@ public virtual void Abort()
                             {
                                 if (NegotiationProtocol.TryParseMessage(buffer, out var negotiationMessage, out consumed, out examined))
                                 {
-                                    var protocol = protocolResolver.GetProtocol(negotiationMessage.Protocol, supportedProtocols, this);
+                                    Protocol = protocolResolver.GetProtocol(negotiationMessage.Protocol, supportedProtocols, this);
 
                                     var transportCapabilities = Features.Get<IConnectionTransportFeature>()?.TransportCapabilities
                                         ?? throw new InvalidOperationException("Unable to read transport capabilities.");
 
-                                    var dataEncoder = (protocol.Type == ProtocolType.Binary && (transportCapabilities & TransferMode.Binary) == 0)
-                                        ? (IDataEncoder)Base64Encoder
-                                        : PassThroughEncoder;
+                                    var requiredTransferMode = (Protocol.Type == ProtocolType.Binary) ?
+                                        TransferFormat.Binary :
+                                        TransferFormat.Text;
+
+                                    if((transportCapabilities & requiredTransferMode) == 0)
+                                    {
+                                        throw new InvalidOperationException($"Cannot use the '{Protocol.Name}' protocol on the current transport. The transport does not support the '{requiredTransferMode}' transfer mode.");
+                                    }
 
                                     var transferModeFeature = Features.Get<ITransferModeFeature>() ??
                                         throw new InvalidOperationException("Unable to read transfer mode.");
 
-                                    transferModeFeature.TransferMode =
-                                        (protocol.Type == ProtocolType.Binary && (transportCapabilities & TransferMode.Binary) != 0)
-                                            ? TransferMode.Binary
-                                            : TransferMode.Text;
+                                    transferModeFeature.TransferMode = requiredTransferMode;
 
-                                    ProtocolReaderWriter = new HubProtocolReaderWriter(protocol, dataEncoder);
-                                    _cachedPingMessage = ProtocolReaderWriter.WriteMessage(PingMessage.Instance);
+                                    using (var ms = new MemoryStream())
+                                    {
+                                        Protocol.WriteMessage(PingMessage.Instance, ms);
+                                        _cachedPingMessage = ms.ToArray();
+                                    }
```

Comment:
What about an internal extension method for WriteMessage on protocol. Newing up MemoryStream is happening a lot when calling this method. Might make refactoring and improving the code simpler if it is in one place.

Reply:
Yes let’s add the helper back.

Reply:
![https://media0.giphy.com/media/l41lZxzroU33typuU/giphy.gif](https://media0.giphy.com/media/l41lZxzroU33typuU/giphy.gif)